### PR TITLE
Misc changes

### PR DESCRIPTION
**Pull request by est77**
- Address basic CUDA classes PR comments.
- Bring back energy compensation tables.
- Small changes to CopyOnWrite class.
- Arrays code cleanup and API improvements.

### CODE REVIEW

File: src/appleseed/renderer/modeling/bsdf/energycompensation.cpp [hunk 1]
```diff
@@ -129,6 +129,15 @@ AlbedoTable2D::AlbedoTable2D()
     m_avg_table = m_albedo_table + TableSize * TableSize;
 }
 
+AlbedoTable2D::AlbedoTable2D(const float* table)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/energycompensation.cpp]: Why are we passing a `const float*` if we need to `const_cast<>` it away immediately after?

File: src/appleseed/foundation/array/array.cpp [hunk 2]
```diff
@@ -44,7 +44,109 @@ namespace foundation
 // Array class implementation.
 //
 
-Array::Concept::~Concept()
+template <typename T>
+Array::Concept* Array::Model<T>::copy() const
```
Comment (dictoon) [inline src/appleseed/foundation/array/array.cpp]: Mark these methods as inline? Not sure it makes much of a difference...

File: src/appleseed/renderer/modeling/bsdf/energycompensation.cpp (hunk 1 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/energycompensation.cpp]: The array is a const float* but to make this class work with both external tables and computed tables I need the const cast

File: src/appleseed/foundation/array/array.cpp (hunk 2 above)
Comment (est77) [reply src/appleseed/foundation/array/array.cpp]: Probably not. All these methods are virtual.

**Pull request merged**